xapian-maintainer-tools/profiling/profiler.c

```diff
@@ -0,0 +1,190 @@
+/** @file profiler.c
+ * @brief C file implementing IO logging
+ */
+/* Compile the profiler with:
+ * gcc -shared -fPIC -o profiler.so profiler.c -ldl
+ *
+ * To use it with a process, use:
+ * LD_PRELOAD=/path/to/profiler.so /path/to/executable
+ *
+ * Running this library will produce an strace.log file,
+ * which can be fed to strace-analyse to produce a log, by running:
+ * ./strace-analyse strace.log
+ */
+
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+
+#define MAX_CALL_LEN 5000
+
+// function for logging calls
+
+int reset = 0;
+
+void logcall(const char *call)
+{
+    FILE *file_ptr;
+    if (!reset) {
+	file_ptr = fopen("./strace.log", "w");
+	reset = 1;
+    } else {
+	file_ptr = fopen("./strace.log", "a");
+    }
+    fprintf(file_ptr, "%s", call);
+    fclose(file_ptr);
+}
+
+// wrapper for open()
+
+typedef int (*real_open_t)(const char *, int, ...);
+
+int open(const char *pathname, int flags, ...)
+{
+    int fd;
+    if (flags & (O_CREAT | O_TMPFILE)) {
+	va_list args_ptr;
+	va_start(args_ptr, flags);
+	fd =
+	    ((real_open_t)dlsym(RTLD_NEXT, "open"))
+	    (pathname, flags, va_arg(args_ptr, mode_t));
+        va_end(args_ptr);
+    } else {
+	fd = ((real_open_t)dlsym(RTLD_NEXT, "open"))(pathname, flags);
+    }
+    char call[MAX_CALL_LEN];
+    char *abspath = realpath(pathname, NULL);
+    snprintf(call, MAX_CALL_LEN, "open(\"%s\",,) = %d\n", abspath, fd);
+    free(abspath);
+    logcall(call);
+    return fd;
+}
+
+// wrapper for close()
+
+typedef int (*real_close_t)(int);
+
+int close(int fd)
+{
+    int return_val = ((real_close_t)dlsym(RTLD_NEXT, "close"))(fd);
+    char call[MAX_CALL_LEN];
+    snprintf(call, MAX_CALL_LEN, "close(%d) = %d\n", fd, return_val);
+    logcall(call);
```

Comment:
Printing the log message to a temporary buffer and then writing the buffer means that we have to impose a fixed limit of the length of log messages.

If you make the `logcall` function take varargs you can just call `logcall("close(%d) = %d\n", fd, return_val);` here, and then in the implementation of `logcall` call `vfprintf()` to pass on the parameters.

Reply:
I'll fix it.